daemon/containerd/image_pull.go

```diff
@@ -68,6 +70,22 @@ func (i *ImageService) PullImage(ctx context.Context, image, tagOrDigest string,
 	infoHandler := snapshotters.AppendInfoHandlerWrapper(ref.String())
 	opts = append(opts, containerd.WithImageHandlerWrapper(infoHandler))
 
-	_, err = i.client.Pull(ctx, ref.String(), opts...)
-	return err
+	img, err := i.client.Pull(ctx, ref.String(), opts...)
+	if err != nil {
+		return err
+	}
+
+	logger := logrus.WithFields(logrus.Fields{
+		"digest": img.Target().Digest,
+		"remote": ref.String(),
+	})
+	logger.Info("image pulled")
+
+	// The pull succeeded, so try to remove any dangling image we have for this target
+	if err := i.client.ImageService().Delete(context.Background(), danglingImageName(img.Target().Digest)); err != nil {
+		if !cerrdefs.IsNotFound(err) {
```

Comment:
🙈 nit: as we already had an `err`, and to _slightly_ reduce cyclo; 

```suggestion
	err = i.client.ImageService().Delete(context.Background(), danglingImageName(img.Target().Digest))
	if err != nil && !cerrdefs.IsNotFound(err) {
```

Perhaps also include the comment from my earlier suggestion to explain why we're ignoring the error here (instead of returning it); https://github.com/moby/moby/pull/45431#discussion_r1180520134

```go
err = i.client.ImageService().Delete(context.Background(), danglingImage.Name)
if err != nil && !cerrdefs.IsNotFound(err) {
	// Image pull succeeded, but cleaning up the dangling image failed. Ignore the
	// error to not mark the pull as failed.
	logger.WithError(err).Warn("unexpected error when deleting dangling image")
}
```

Reply:
Ahh yeah, I'll pull those comments in and change the wording with @tianon's suggestion, I think that's the cleanest.